python/ray/serve/backend_state_manager.py

```diff
@@ -948,23 +1039,55 @@ def _completed_goals(self) -> List[GoalId]:
             ]) > 0):
                 continue
 
-            running = self._replicas[backend_tag].get(
+            # All replicas are in steady or terminal states beyond this point
+            # thus we can start tracking completed goals.
+            target_version = self._target_versions[backend_tag]
+
+            all_running_replica = self._replicas[backend_tag].get(
                 states=[ReplicaState.RUNNING])
-            running_count = len(running)
+            running_at_target_version_replica = [
+                replica.version == target_version
+                for replica in all_running_replica
+            ]
+            failed_to_start_replica = self._replicas[backend_tag].get(
+                states=[ReplicaState.FAILED_TO_START])
 
             # Check for deleting.
-            if target_count == 0 and running_count == 0:
+            if target_replica_count == 0 and len(all_running_replica) == 0:
                 deleted_backends.append(backend_tag)
                 completed_goals.append(
                     self._backend_goals.pop(backend_tag, None))
 
             # Check for a non-zero number of backends.
-            elif target_count == running_count:
-                # Check that all running replicas are the target version.
-                target_version = self._target_versions[backend_tag]
-                if all(r.version == target_version for r in running):
-                    completed_goals.append(
-                        self._backend_goals.pop(backend_tag, None))
+            elif target_replica_count == len(
+                    running_at_target_version_replica):
+                completed_goals.append(
+                    self._backend_goals.pop(backend_tag, None))
+
+            elif len(failed_to_start_replica) > 0 and len(
+                    running_at_target_version_replica) > 0:
+                # Reconfigure action is considered partially successful since
+                # we had at least one running replica at target version, thus
+                # let user code return but serve controller will keep trying
+                # to bring to target running replica fully
```

Comment:
I think we should only return once the desired number of replicas is reached, this would be a behavior change.

Reply:
Currently the behavior is "return once the desired state is reached" which is much more intuitive